### DISCUSSION TITLE
`ros2 topic list` not displaying in Ubuntu 22.04

### DISCUSSION THREAD

**Discussion by iansexton1** [iRobotEducation/create3_docs#211] [Questions]
I just downloaded Ubuntu 22.04 and going through the setup under "Install ROS 2 Humble with Create 3 Messages on an Ubuntu 22.04 Machine". I went through all the steps but once I got to the step where I typed "ros2 topic list" the only thing that returns is /parameter_events
/rosout
Im able to connect to the robot on the Python interface as well as ping its IP address in ubuntu but can't run any operations for the robot

**Comment by alsora** [comment]
Hi @iansexton1!
First of all an important clarification: ROS 2 does not guarantee communication between different distributions (e.g. an application running ROS 2 Galactic may have problems communicating with another application running ROS 2 Humble).
Your robot is running ROS 2 Galactic (we haven't released a firmware with ROS 2 Humble yet, this will happen soon and I recommend subscribing to this discussion to be notified of the release https://github.com/iRobotEducation/create3_docs/discussions/70).
The firmware versions have a prefix denoting the ROS 2 distribution: e.g. `G.3.1` uses ROS 2 `G`alactic.

Sorry for the long premise, but I wanted to make it clear as not following it may lead to hard to debug problems.
We can still try to get your setup working, but if the problems persist I highly recommend to try again using matching ROS 2 distributions.

Besides the ROS 2 distribution, another critical configuration that affects ROS 2 remote communication is the choice of the ROS 2 Middleware (RMW).
You need to make sure that applications are using the same RMW to allow communication.
Moreover, note that different ROS 2 distributions have different default RMWs: the default for ROS 2 Galactic is `rmw_cyclonedds_cpp` while the default for ROS 2 Humble is `rmw_fastrtps_cpp`.

Thus, unless you manually modified the default RMWs, your setup is currently using two incompatible middlewares.
I recommend to set the RMW to `rmw_fastrtps_cpp` for both machines.

On your Ubuntu 22.04 setup you can do the following
```
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
```

For the robot, you should connect to the webserver, go into the "Application" tab, select `rmw_fastrtps_cpp` from the RMW_IMPLEMENTATION drop down menu and then reboot the robot application (see https://iroboteducation.github.io/create3_docs/webserver/application/).

After you have set both RMWs to the same value, please try again to list all the topics using the following command:
```
ros2 topic list --spin-time 10 --no-daemon
```
The additional arguments `--spin-time 10` is used to spend more time searching for the robot and `--no-daemon` is used for running discovery from scratch without relying on cached information.

I hope this fixes your problem!
Let me know the results!

**Reply by shamlian** [reply]
Checking in -- any luck, @iansexton1 ?

**Comment by swilloughbypme** [comment]
Tried the H.0.0 firmware update, and I'm still having the same issue as the OP.  The only topics listed are /parameter_events and /rosout. 

Notes:
- I'm using a VirtualBox VM running 22.04 Jammy
- Humble is installed and working
- I ensured both the robot and the VM are both using Fast DDS  (printenv RMW_IMPLEMENTATION yields rmw_fastrtps_cpp on the VM) 
- Both are on the same wifi network, which is my home network
- I used the recommended command:  ros2 topic list --spin-time 10 --no-daemon
- The robot can be pinged successfully from the VM by ip address 
- I've restarted the VM and robot a couple of times (yep, desperation has set in.)
- In the robot's app config page, the domain id is 0 and there is no namespace entered

One thing I did notice was that after rebooting the VM, the RMW was set back to cyclone. I changed the robot to also use cyclone, and still could only list the same 2 topics.

**Reply by alsora** [reply]
Virtual machines are not great with ROS 2 if you need to communicate remotely.
Depending on the particular virtual machine that you are using, you may have to configure it such that it can freely access your network interfaces.

Unfortunately, the fact that you can ping the robot is not sufficient to guarantee ROS communication. 

Can you try to use a native installation of ROS2 rather than a vm? 

Unless you save the RMW in a configuration file such as ~/.bashrc, everytime you restart the vm it will go back to the default (cyclonedds for galactic or fastdds for humble). 

The problem may also be in your router/firewall or laptop network interface (they all may be configured to block multicast UDP packets). 
You may want to try to configure fastdds using an xml profile that disables multicast, as explained here 
https://iroboteducation.github.io/create3_docs/setup/xml-config/

**Reply by iansexton1** [reply]
Is there a native installation of ROS 2 that you'd recommend? I've tried downloading Ubuntu 22.04 from the Windows store but experiencing the same issue.

 In an earlier reply, you mentioned that the robot is running ROS 2 Galactic (we haven't released a firmware with ROS 2 Humble yet). Wouldnt this mean that we would be unable to use Ubuntu 22.04 with the create 3 as 22.04 is meant for ROS 2 Humble?

**Reply by alsora** [reply]
First of all, I think that what you download/install from the windows store allows you to use Ubuntu in the windows subsystem for Linux (WSL), which as mentioned by @retupmoc1 in this discussion, does not work out of the box.

A "native installation" requires that you install an Ubuntu operative system on your computer, e.g. alongside windows. 
There are plenty of guides online if you search for it

If you plan to use ROS 2 galactic, then I recommend installing Ubuntu 20.04.
Here you can then Install ROS from Debian packages in a single command.

You can still use ROS 2 galactic on successive versions of Ubuntu (e.g. 22.04), but it will have to be built from sources which I only recommend to expert users.

**Reply by swilloughbypme** [reply]
Just a quick clarification, I'm using an Oracle VirtualBox VM rather than WSL.

There were 2 things I had to do to make the VM work with the Create3.

- In the VM's Network Settings, I changed "Attached to" from NAT to "Bridged" and "Promiscuous Mode" to "Allow All." Then I could list the topics on the Create3, but the example commands failed with this error: "The passed action type is invalid"
-  The solution for this was to install the Create3 messages on the VM as noted [here](https://github.com/iRobotEducation/create3_docs/discussions/178). This has to be done regardless of whether you use a VM or a native Ubuntu installation. 

FYI, there is a Humble Create3 firmware beta release (H.0.0) that works on Create3 if you don't need the BLE functionality. It's noted [here](https://github.com/iRobotEducation/create3_docs/discussions/70). Then you can use Humble on an Ubuntu 22.04 VM or native install if you prefer Humble over Galactic. I've been this configuration for a few days with no issues so far, but it's a beta release, so proceed with caution. I don't believe you can go easily go back to a G.x.x firmware version after installing H.0.0.

Also, the messages apk for Humble is ros-**humble**-irobot-create-msgs rather than ros-**galactic**-irobot-create-msgs.

Hope this helps and thanks again to @alsora and others who have been so knowledgeable and quick to respond to questions!
Steve

**Comment by swilloughbypme** [comment]
Hi Alberto,

Great info. I'll try all try all your suggestions.

Thanks!
Steve
